### ISSUE TITLE
Parallel Task and Exception handling fixes

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#888]
This issue was originally created at: 2004-08-22 23:21:53.
This issue was reported by: `kquick`.
kquick said at 2004-08-22 23:21:53
>Summary: Task scheduling and exception handling fixes.
> 
>    Existing exception handling had some race conditions and other 
>issues,
>    exhibited in the following ways:
>    
>    * Occasional failures of test/option-j.py with a file-not-found error
>      for /.../f4.  This file should have been created by the second
>      thread of a -j2 test (where the first build for f3 returned an error
>      causing an exception).  The start() method of SCons.Job.Parallel
>      would start looking for task completions as soon as a task was
>      issued; sometimes it would see the exception completion of the first
>      task before scheduling the second task, making results of this test
>      indeterminate.
>    
>    * Occasional failures of test/option-j.py with two "scons: *** [f3]
>      Error 1" messages instead of the expected single message.  This is
>      caused by the fact that a Task level exception was being set
>      globally in the Taskmaster, regardless of what the Taskmaster was
>      doing.  Thus, the Taskmaster could be preparing the Task for the 
>f4
>      build when the f3 build set the Taskmaster-global error.
>    
>    * Related to the above, exceptions were signalled from the 
>Taskmaster
>      rather than from the task level, so often (in the Parallel case) a
>      task's exception would actually be signalled when preparing (or
>      handling) and associated with the following task, or even much later
>      tasks if the exception is not immediately generated.
>    
>    Although most of the above were benign, they caused indeterminate
>    results, violated thread cleanliness, could produce misleading
>    exception information, and made inefficient use of thread resources.
>    The following changes were made to address these issues.
>    
>    * Job.Parallel.start() was modified to try to issue numjobs at any one
>      time.  At the point that start() is invoked, all threads have
>      already been created, so a width-first dispersal of Tasks is more
>      efficient than the depth-first method previously utilized, and it
>      also establishes a (fairly) deterministic execution of Tasks:
>    
>       A) At least numjobs of Tasks will be attempted (where numjobs is
>          the -j N value, and assuming that at least that many Tasks
>          exist).
>       B) If a task encounters an exception, that exception will (in the
>          default fail_stop() configuration) block any further scheduling
>          of Tasks, but it will not have any effect on the Tasks currently
>          executing on other threads (barring grand-mal shutdown by
>          Task/Taskmaster subclassed instances).
>       C) Each time a thread completes a Task, another Task will be
>          scheduled unless blocked by a previous exception.
>    
>    * The Taskmaster retains a raise_exception() method that can be 
>used
>      to define specific handling for Task exceptions; the default
>      handling is to propagate the exception.  All other per-task
>      exception handling is moved to the Task object instead.  
>Exceptions
>      encountered in next_task() and task.prepare() are attached to the
>      task when created (via Taskmaster.ready_exc, maintained in 
>parallel
>      to the Taskmaster.ready node).
> 
>    * Due to the above changes, all exception-driven diagnostic output 
>for
>      *all* simultaneous tasks is reported rather than exception output
>      from only one task.  This is verified by additional tests in
>      test/exceptions.py.
>    
>    * The Job.Threadpool object adds a preparation_failed() method 
>which
>      places a task which fails during prepare() (or
>      Taskmaster.next_task()) directly onto the resultQueue as a failure.
>      This preserves ordering of handling and fixes a bug wherein the
>      Task.failed() method could be called twice for a single Task.
>    
>    * The Job.Threadpool no longer needs the get_nowait method and 
>the
>      start() jobs counter becomes method-local instead of an instance
>      variable.
> 
>    * The JobTests.py tests were expanded to perform a series of tests
>      using the SCons.Task.Task and SCons.Task.Taskmaster objects 
>with the
>      Job objects (in addition to the internal contrived Task and
>      Taskmaster objects tested to-date).  This validates both textbook
>      and integrated functionality of the Job and Task/Taskmaster
>      operations.  There's some rather ugly logic in the tests that
>      attempts to ensure that the expected sequence of Task 
>completions,
>      failures, and non-executions occurs; this logic assumes that a
>      "time.sleep(...)" operation will allow alternate threads to execute.
>    
>    * Updated test/exceptions.py to test exceptions generated by builder
>      action return values as well as those exceptions raised directly in
>      the builder action.

stevenknight said at 2004-08-31 04:00:29
>Hey Kevin--
> 
>Wow!  Thanks for all of the hard work on this one,
>especially the test cases.
> 
>Unfortunately, I'm getting a prettyconsistent failure from
>JobTests.py when run under Python 1.5.2 on my Red Hat 9
>system.  I tried it about six times, and it did succeed
>once, but it failed all of the other times.  I've attached a
>file with the failure output so you can take a look.
> 
>I'm not very thread-savvy, so let me know what additional
>information I can provide to help track this down.  Good
>parallel execution, I think, one of SCons' biggest selling
>points, so I'm really eager to fix whatever corner case this
>is and get this patch integrated.
> 
>Thanks!
> 
>        --SK

mathewboorman said at 2004-09-09 07:37:12
>The failing test case uses sleeps to create the appropriate
>order of the events.  This is always going to be a problem
>as CPU speed and scheduling variability comes into play. 
>Also it means the unit tests take much longer to run as they
>spend alot of time just sleeping.
> 
>I suggest instead modifing the custom Nodes so they
>synchronise with each other to get the desired build order.  
> 
>A rough way this may be able to work is:
>The test thread to creates a 'clock' variable and use
>threading.Condition to protect it.  Pass these both to all
>the nodes.
>The Nodes run a loop that aquires the clock, checks if they
>are scheduled  to finish, does a notify again, and then exit
>if appropriate.
>The test thread also has a loop that acquires the clock, and
>increments it.

kquick said at 2004-09-09 21:35:44
>Shamefully, I had arrived at the same conclusion regarding the failing test.  In 
>the case that Steven posted, lengthening the sleep times made the problems 
>go away (assuming either a CPU speed issue or a sleep resolution issue).  
>However, a multi-processor machine would also fail.
> 
>I spent some time considering how to try to resolve this issue.  I had 
>originally gotten caught up in this test trying to justify the change in the Jobs.
>py Parallel scheduling loop that I included in this patch; unfortunately I got 
>too focused on details... ergo the naive test and subsequent issues.  I like 
>Mathew's proposal, and spent some time trying to determine if something like 
>that would work, but came to the conclusing that it's not really a meaningful 
>test at that point (also there's an issue that the test doesn't have control while 
>the threads are running, so the test would have to create yet another thread 
>that did the clock incrementing).  I decided that the result would only really 
>show that multiple threads were running, which was already proven earlier in 
>the JobTests.py. 
> 
>End result is that I greatly simplified this test to ensure that at least one of 
>the jobs ran and that each test node either didn't execute or ended up in the 
>expected success or failure state.  Anything more would require enforced 
>determinism like Mathew's proposal, or complete knowledge of thread 
>scheduling for the system.  And as I said, I'm not sure the former gives any 
>meaningful test indication and the latter is unrealistic/impossible.  The test 
>can still (and does) ensure that none of the crosstalk problems with 
>exceptions and completion status information occurs.
> 
>End result is that the problems appear to be unrealistic expectations in the 
>test and that the patch to SCons itself is solid.  The scheduling change can't 
>realistically be tested, other than its ability to reasonably schedule threads, 
>although I still stand behind my modification there because I think it increases 
>the parallelism and makes more effective use of the already instantiated 
>threads.
> 
>I have communicated this to Steven, and queried him about an additional issue 
>he noted, but he's apparently out-of-pocket at the present time, so I'm just 
>waiting for his response.
> 
>Of course, I'd welcome any enhancements to the tests if anyone were so 
>motivated! :)

stevenknight said at 2004-09-17 05:55:30
>Integrated for next release.  Thank you!  --SK

issues@scons said at 2004-09-17 05:55:30
>Converted from SourceForge tracker item 1014089
